### PR TITLE
use upsert when update entities

### PR DESCRIPTION
**Pull request by AlvaroVega**
issue https://github.com/telefonicaid/iotagent-node-lib/issues/956

This will be enable to not create initially entities (actually is done even for autorpovisioned devices)

### CODE REVIEW

File: lib/services/devices/devices-NGSI-v2.js [hunk 1]
```diff
@@ -269,7 +269,7 @@ function createInitialEntityNgsi2(deviceData, newDevice, callback) {
  */
 function updateEntityNgsi2(deviceData, updatedDevice, callback) {
```
Comment (fgalan) [inline lib/services/devices/devices-NGSI-v2.js:270]: If the same CB API method is used at the end for both create and update cases, why do we need createInitialEntityNgsi2() and updateEntityNgsi2() separate functions? Maybe the can be unified? Or is some additional logic beyond CB request specific in each case?

Comment (AlvaroVega) [reply lib/services/devices/devices-NGSI-v2.js:270]: Now both functions are very similar, but it seems they are not following the same logic

Comment (fgalan) [reply lib/services/devices/devices-NGSI-v2.js:270]: In this PR on in a later one?

Comment (AlvaroVega) [reply lib/services/devices/devices-NGSI-v2.js:270]: IMHO in another PR

Comment (fgalan) [reply lib/services/devices/devices-NGSI-v2.js:270]: NTC

**Review by fgalan — APPROVED** [review]
LGTM

**Comment by AlvaroVega** [comment]
Currently autorpovision is create initial entity and then performs an update. If we add this PR we can achieve in the future that autorpovision just update an entity without create it initially (see https://github.com/telefonicaid/iotagent-node-lib/issues/1334)

**Comment by fgalan** [comment]
> Currently autorpovision is create initial entity and then performs an update. If we add this PR we can achieve in the future that autorpovision just update an entity without create it initially (see #1334)

@manucarrace what do you think?

**Comment by fgalan** [comment]
This PR would make `appendMode` configuration setting useless. It would be as hardwiring this setting to always `true`. We would gain in simplicity (just one way to interact with CB) but we would loss flexibility.

**Comment by AlvaroVega** [comment]
@fgalan how is related upsert with  appendMode of CB update?

https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/installationguide.md

appendMode: if this flag is activated, the update requests to the Context Broker will be performed always with APPEND type, instead of the default UPDATE. This have implications in the use of attributes with Context Providers, so this flag should be used with care. This flag is overwritten by autoprovision flag in group or device provision.

**Comment by AlvaroVega** [comment]
https://github.com/telefonicaid/iotagent-node-lib/blob/ad90059d58a6c85b282547812dd0cd7ee01b97b6/lib/services/ngsi/entities-NGSI-v2.js#L351-L361

File: lib/services/ngsi/entities-NGSI-v2.js [hunk 2]
```diff
@@ -970,16 +970,19 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
...
             // recreate options object to use single entity update
-            url = '/v2/entities/' + entityName + '/attrs';
-            if (typeInformation && typeInformation.type) {
-                url += '?type=' + typeInformation.type;
+            url = '/v2/entities?options=upsert';
+            if (config.getConfig().appendMode === false) {
+                url = '/v2/entities/' + entityName + '/attrs';
+                if (typeInformation && typeInformation.type) {
+                    url += '?type=' + typeInformation.type;
+                }
```
Comment (fgalan) [inline lib/services/ngsi/entities-NGSI-v2.js:978]:
```suggestion
            url = '/v2/entities';
            if (config.getConfig().appendMode === false) {
                if (typeInformation && typeInformation.type) {
                    url += '?type=' + typeInformation.type;
                }
            }
            else {  // false
                url += '?options=upsert
            }
```

File: lib/services/ngsi/entities-NGSI-v2.js [hunk 3]
```diff
@@ -970,16 +970,22 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
...
         if (options.json.entities.length === 1) {
             // recreate options object to use single entity update
-            url = '/v2/entities/' + entityName + '/attrs';
-            if (typeInformation && typeInformation.type) {
-                url += '?type=' + typeInformation.type;
+            url = '/v2/entities';
+            if (config.getConfig().appendMode === false) {
+                url += '/' + entityName + '/attrs';
+                if (typeInformation && typeInformation.type) {
+                    url += +'?type=' + typeInformation.type;
```
Comment (fgalan) [inline lib/services/ngsi/entities-NGSI-v2.js:977]: I think this should be:

```suggestion
                    url += options.json.id +'/attrs?type=' + typeInformation.type;
```

(Not sure if `options.json.id` is the right variable here to get entity id)

File: lib/services/devices/devices-NGSI-v2.js [hunk 4]
```diff
@@ -265,7 +265,7 @@ function createInitialEntityNgsi2(deviceData, newDevice, callback) {
...
     const options = {
-        url: config.getConfig().contextBroker.url + '/v2/entities/' + String(deviceData.name) + '/attrs',
+        url: config.getConfig().contextBroker.url + '/v2/entities/' + String(deviceData.name) + '/attrs?options=upsert',
```
Comment (fgalan) [inline lib/services/devices/devices-NGSI-v2.js:268]: This operation is not going to work... CB only accepts `?options=upsert` in `POST /v2/entities` not in `POST /v2/entities/{id}/attrs`...

In addition, maybe `if (config.getConfig().appendMode === ...)` should be used also here?

File: lib/services/devices/devices-NGSI-v2.js [hunk 5]
```diff
@@ -276,13 +276,14 @@ function updateEntityNgsi2(deviceData, updatedDevice, callback) {
     };
 
     if (deviceData.cbHost && deviceData.cbHost.indexOf('://') !== -1) {
-        options.url = deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs';
+        options.url = deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs?options=upsert';
     } else if (deviceData.cbHost && deviceData.cbHost.indexOf('://') === -1) {
-        options.url = 'http://' + deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs';
+        options.url =
+            'http://' + deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs?options=upsert';
     }
```
Comment (fgalan) [inline lib/services/devices/devices-NGSI-v2.js:283]: Similar comment to the one above.

File: lib/services/devices/devices-NGSI-v2.js (hunk 4 above)
Comment (AlvaroVega) [reply lib/services/devices/devices-NGSI-v2.js]: Then I'll remove `?options=upsert` for that case. That case ins not the usual flow. An check apppendMode is not expected here

Comment (AlvaroVega) [reply lib/services/devices/devices-NGSI-v2.js]: That flow is for update a register, and has not being modified; this PR just use upsert when update an entity after get a measure

File: CHANGES_NEXT_RELEASE [hunk 6]
```diff
@@ -1 +1,2 @@
+- Fix: use 'options=upsert' when update ngsiv2 CB entities (#956)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
- Fix: use 'options=upsert' when update ngsiv2 CB entities and appendMode is enabled (#956)
```

File: lib/services/ngsi/entities-NGSI-v2.js [hunk 7]
```diff
@@ -970,16 +970,22 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
...
-            if (typeInformation && typeInformation.type) {
-                url += '?type=' + typeInformation.type;
+            url = '/v2/entities';
+            if (config.getConfig().appendMode === false) {
+                url += '/' + entityName + '/attrs';
+                if (typeInformation && typeInformation.type) {
+                    url += '?type=' + typeInformation.type;
+                }
+            } else {
+                // false
```
Comment (fgalan) [inline lib/services/ngsi/entities-NGSI-v2.js:980]:
```suggestion
                // appendMode === true
```

File: test/unit/ngsiv2/ngsiService/active-devices-test.js [hunk 8]
```diff
@@ -193,26 +192,27 @@ describe('NGSI-v2 - Active attributes test', function () {
...
+        //                 value: 'wrongid'
+        //             }
+        //         ];
+
+        //         iotAgentLib.update('light1', 'Light', '', wrongvalues.concat(values), function (error) {
+        //             should.not.exist(error);
+        //             contextBrokerMock.done();
+        //             done();
+        //         });
+        //     });
```
Comment (fgalan) [inline test/unit/ngsiv2/ngsiService/active-devices-test.js:215]: Should be re-enabled?

File: test/unit/ngsiv2/expressions/jexlBasedTransformations-test.js [hunk 9]
```diff
@@ -1123,12 +1107,12 @@ describe('Java expression language (JEXL) based transformations plugin', functio
...
                 .matchHeader('fiware-servicepath', 'gardens')
                 .post(
-                    '/v2/entities/gps1/attrs',
+                    '/v2/entities?options=upsert',
                     utils.readExampleFile(
                         './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin32.json'
                     )
                 )
-                .query({ type: 'GPS' })
+                //uery({ type: 'GPS' })
```
Comment (fgalan) [inline test/unit/ngsiv2/expressions/jexlBasedTransformations-test.js:1115]: Lefotver?

Comment (AlvaroVega) [reply test/unit/ngsiv2/expressions/jexlBasedTransformations-test.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1111/commits/2fb5b110a37aacba924480c621934963462df562

**Review by fgalan — APPROVED** [review]
LGTM

Pendiong on @mapedraza review and LGTM

**Review by mapedraza — APPROVED** [review]
LGTM

**Pull request merged**